qualtran/bloqs/mod_arithmetic/mod_subtraction_test.py

```diff
@@ -0,0 +1,115 @@
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 10)]
+)
+def test_valid_modneg_decomposition(dtype, bitsize, prime):
+    b = ModNeg(dtype(bitsize), prime)
+    qlt_testing.assert_valid_bloq_decomposition(b)
+    qlt_testing.assert_equivalent_bloq_counts(b, [ignore_split_join, ignore_alloc_free])
+
+
+@pytest.mark.parametrize('cv', range(2))
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 10)]
+)
+def test_valid_cmodneg_decomposition(dtype, bitsize, prime, cv):
+    b = CModNeg(dtype(bitsize), prime, cv)
+    qlt_testing.assert_valid_bloq_decomposition(b)
+    qlt_testing.assert_equivalent_bloq_counts(b, [ignore_split_join, ignore_alloc_free])
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 10)]
+)
+def test_modneg_classical_action(dtype, bitsize, prime):
+    b = ModNeg(dtype(bitsize), prime)
+    cb = b.decompose_bloq()
+    for x in range(prime):
+        assert b.call_classically(x=x) == cb.call_classically(x=x) == ((-x) % prime,)
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize('cv', range(2))
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['bitsize', 'prime'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 10)]
+)
+def test_cmodneg_classical_action(dtype, bitsize, prime, cv):
+    b = CModNeg(dtype(bitsize), prime, cv)
+    cb = b.decompose_bloq()
+    for control in range(2):
+        for x in range(prime):
+            assert b.call_classically(ctrl=control, x=x) == cb.call_classically(ctrl=control, x=x)
+
+
+def test_modneg_classical_action_action_fast():
+    prime = 11
+    b = ModNeg(QMontgomeryUInt(4), prime)
+    cb = b.decompose_bloq()
+    for x in range(prime):
+        assert b.call_classically(x=x) == cb.call_classically(x=x) == ((-x) % prime,)
+
+
+@pytest.mark.parametrize('cv', range(2))
+def test_cmodneg_classical_action_fast(cv):
+    prime = 11
+    b = CModNeg(QMontgomeryUInt(4), prime, cv)
+    cb = b.decompose_bloq()
+    for control in range(2):
+        for x in range(prime):
+            assert b.call_classically(ctrl=control, x=x) == cb.call_classically(ctrl=control, x=x)
+
+
+def test_modneg_cost():
+    n, p = sympy.symbols('n p')
+    b = ModNeg(QMontgomeryUInt(n), p)
+    counts = get_cost_value(b, QECGatesCost()).total_t_and_ccz_count()
+    assert counts['n_t'] == 0, 'all toffoli'
+    assert counts['n_ccz'] == 3 * (n - 1)
+
+
+def test_cmodneg_cost():
+    n, p = sympy.symbols('n p')
+    for cv in range(2):
+        b = CModNeg(QMontgomeryUInt(n), p, cv)
+        counts = get_cost_value(b, QECGatesCost()).total_t_and_ccz_count()
+
+        assert counts['n_t'] == 0, 'all toffoli'
+        assert counts['n_ccz'] == 3 * (n - 1) + 1
```

Comment:
add a note/comment with litinski fig/table 8

Reply:
I didn't notice this at first but I'm not using the controlled decomposition from litinski ... this is my own decomposition which matches the cost but for the wrong reason... they turn the CNOTs into toffolis which is not necessary. I just combine the old control qubit and outer control qubit into an ancilla qubit
![image](https://github.com/user-attachments/assets/1a2a7af5-5174-4bd5-b73f-6c7e80013563)


---

for the uncontrolled version we have $3n$ instead of $2n$ because we are not the measurement uncomputation part because as we discussed the measurement thing will add random phase flips.

Reply:
>  this is my own decomposition

This is important to capture somewhere in the docs

Reply:
done